iree/compiler/InputConversion/Common/ConvertUpstreamToIREE.cpp

```diff
@@ -0,0 +1,131 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Flow/IR/FlowDialect.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/InputConversion/Common/PassDetail.h"
+#include "iree/compiler/InputConversion/Common/Passes.h"
+#include "mlir/Dialect/Linalg/IR/LinalgOps.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+
+struct ConvertTensorCastPattern : public OpConversionPattern<tensor::CastOp> {
+  using OpConversionPattern<tensor::CastOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      tensor::CastOp op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+    Value input = operands.front();
+    ShapedType inputType = input.getType().dyn_cast<ShapedType>();
+    ShapedType resultType =
+        typeConverter->convertType(op.getType()).dyn_cast_or_null<ShapedType>();
+    if (!inputType || !resultType || !inputType.hasRank() ||
+        !resultType.hasRank()) {
+      return rewriter.notifyMatchFailure(op, "not ranked shaped types");
+    }
+    // This should not happen, except in the context of type conversion.
+    if (inputType.getRank() != resultType.getRank()) {
+      return rewriter.notifyMatchFailure(op, "mismatched rank");
+    }
+
+    // Resolve dims to the most specific value.
+    int rank = inputType.getRank();
+    SmallVector<Value> dimSizes(rank);
+    auto resolveDimSize = [&](int position) -> Value {
+      if (!dimSizes[position]) {
+        // Find the most specific.
+        if (!inputType.isDynamicDim(position) ||
+            !resultType.isDynamicDim(position)) {
+          // Static dim.
+          int64_t dimSize = !inputType.isDynamicDim(position)
+                                ? inputType.getDimSize(position)
+                                : resultType.getDimSize(position);
+          dimSizes[position] = rewriter.create<ConstantIndexOp>(loc, dimSize);
+        } else {
+          // Dynamic dim.
+          dimSizes[position] =
+              rewriter.create<tensor::DimOp>(loc, input, position);
+        }
+      }
+
+      return dimSizes[position];
+    };
+
+    SmallVector<Value> sourceDynamicDims;
+    SmallVector<Value> targetDynamicDims;
+    for (int i = 0; i < rank; i++) {
+      if (inputType.isDynamicDim(i)) {
+        sourceDynamicDims.push_back(resolveDimSize(i));
+      }
+      if (resultType.isDynamicDim(i)) {
+        targetDynamicDims.push_back(resolveDimSize(i));
+      }
+    }
+
+    rewriter.replaceOpWithNewOp<IREE::Flow::TensorReshapeOp>(
```

Comment:
Not really related to this change, but using `flow.tensor.reshape` here doesnt feel right. It would be better to introduce a `flow.tensor.cast` operation, which just lowers to a `hal.*.cast` operation. I think might lead us down a path where we will then special case this reshape to lower to a cast. Didnt bring this up previously on Stellas PR since she was just going off to vacation, and I dont have the cycles to plumb this through. Could you file a bug, and mention that bug here as a TODO to track this. (If you are up for plumbing this through that would be awesome, but that would probably be more than what you signed up for :) ).

Reply:
hal.tensor.cast is only for ABI boundaries, but here we are handling the general internal case by turning it into a `flow.tensor.reshape` (which I don't think turns into `hal.tensor.cast` -- at the hal level, we only have !hal.buffer, which is shape-erased anyway so the reshape turns into nothing).

Reply:
Thanks for the comments. I opened https://github.com/google/iree/issues/6418 for tracking or further discussion.